### ISSUE TITLE
ETS case Behavior in Queue Mode

### ISSUE THREAD

**Issue by Kucmasz** [OpenMobileAlliance/OMA_LwM2M_for_Developers#584]
In doc MA-ETS-LightweightM2M_INT-V1_1-20190912-D test case LightweightM2M-1.1-int-109 – Behavior in Queue Mode it is said to set lifetime as 240sec to verify queue mode behavior and see whether GET query is handled after update message, but for UDP connection 240seconds actually cause the NAT timeout on the network and client needs to re-register instead of basically updating the registration, as the new source port will be assigned.

**Comment by sbernard31** [comment]
(I'm not an OMA member)

Note that the test says : 
> Server changes the lifetime of the client to some long value (**e.g. 240**), by sending a Write (CoAP POST) on /1/x/1
(where x is LwM2M Server Object Instance) with value of 240

So you can use something smaller than 240 but not sure it will solves your issue because  ... 

Even if [RFC4787](https://www.rfc-editor.org/rfc/rfc4787.html#section-4.3) says : 
> A NAT UDP mapping timer MUST NOT expire in less than two  minutes.
> ...
>  A default value of five minutes or more for the NAT UDP mapping timer is RECOMMENDED.

Some studies shows smaller timeouts in real world: 
![Capture d’écran du 2024-06-11 12-02-06](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/assets/840294/0b262164-6a6a-40a8-bdae-9f4dba522679)

(source : https://www.researchgate.net/publication/221198307_UDP_NAT_and_firewall_puncturing_in_the_wild)

**But,** even if you get NAT timeout, I'm not sure to get why this is an issue ? 
The idea of using Queue mode is (among other) to be able to traverse NAT. 
The spec says : 
> Queue Mode may help with NAT traversal by allowing a LwM2M Server to send messages to the LwM2M Client only after the LwM2M Client has initiated the communication to the LwM2M server first. This ensures that NAT bindings exist that allow incoming packets to the LwM2M Client.

> The LwM2M Client lets the LwM2M Server know it is awake by sending a registration update message

So why sending an UPDATE will/should not work because of NAT if new source port will be assigned ?

**Comment by avs-jgawron** [comment]
That is true, but after NAT Timeout, without DTLS Connection ID, Registration Update will usually be ignored by the server (no way to verify identity especially for nosec connection), which will force client to perform full register. I do not recall any mention in any spec forcing any server to make port-ip obligatory for nosec connection though. For DTLS without Session Resumption / Connection ID it will break the session because of 5-touple.

Just my 2 cents.

**Comment by sbernard31** [comment]
> I do not recall any mention in any spec forcing any server to make port-ip obligatory for nosec connection though.

I didn't remember too but I find [LWM2M-v1.2.1@transport§6.4.3. Registration Interface](https://www.openmobilealliance.org/release/LightweightM2M/V1_2_1-20221209-A/HTML-Version/OMA-TS-LightweightM2M_Transport-V1_2_1-20221209-A.html#6-4-3-0-643-Registration-Interface): 

_"When the LwM2M Client IP address changes in NoSec mode, the Client MUST register again to the LwM2M Server."_ 

> Registration Update will usually be ignored by the server (no way to verify identity especially for nosec connection)
 
My guess is that `nosec` is not really production ready but mainly(only?) for development/debug/test purpose.  In that case, you can imagine that even if IP/port changed then UPDATE can be accepted  (especially as UPDATE request contain registration ID)
(but if you strictly follow the spec you can not :point_up: )

> For DTLS without Session Resumption / Connection ID it will break the session because of 5-tuple mismatch.

Yep and this is another (but interesting) question to know if a new  DTLS session should really need a new registration. (https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/506)
I think last version of the spec doesn't force it anymore.

**Comment by avs-jgawron** [comment]
Thank you for research! It looks like it boils down to 1.1 vs 1.2. In 1.1 You test requirements can be met only for nosec and for 1.2 only for dtls. 

> My guess is that nosec is not really production ready but mainly(only?) for development/debug/test purpose.

I think there are some usecases for nosec running in production, although I'm not a fan of any of them 😄

**Comment by mkgillmore** [comment]
This behavior can be fixed with CID's.  Future version should support CID's
